pmd-ant/src/main/java/net/sourceforge/pmd/ant/Formatter.java

```diff
@@ -210,8 +211,21 @@ private static Writer getToFileWriter(String baseDir, File toFile, Charset chars
 
     private static String getConsoleEncoding() {
         Console console = System.console();
-        // in case of pipe or redirect, no interactive console.
+        // in case of pipe or redirect, no interactive console, we get null
         if (console != null) {
+            // Since Java 22, this returns a console even for redirected streams.
+            // In that case, we need to check Console.isTerminal()
+            // See: JLine As The Default Console Provider (JDK-8308591)
+            try {
+                Boolean isTerminal = (Boolean) MethodUtils.invokeMethod(console, "isTerminal");
+                if (!isTerminal) {
+                    // stop here, we don't have an interactive console.
+                    return null;
+                }
+            } catch (InvocationTargetException | NoSuchMethodException | IllegalAccessException ignored) {
+                // fall-through - we use a Java Runtime < 22.
+            }
+
```

Comment:
I'd assume since our CLI progressbar uses JLine3 under the hood to detect terminal capabilities that it is safe and won't choke on this change… but did you have the chance to test it?

Reply:
This change is about how we determine the console encoding for ant, if the report should be output to the console (`toConsole=true`) - and when building PMD with Java 22, the (ant) unit tests failed, because then we couldn't override the encoding for testing purpose anymore and couldn't test various encoding combinations. The tests are actually in pmd-java:

https://github.com/pmd/pmd/blob/04c2c0106ddbd200fb8a978849606389757dd7c0/pmd-java/src/test/java/net/sourceforge/pmd/ant/PMDTaskTest.java#L89-L110

And we explicitly set the system property "file.encoding" in the unit test, so that we can "override" the encoding for the test. That didn't work anymore, because Java 22 now returns always a Console and the tests used then always the correct console encoding rather than the faked test encoding.

I actually didn't test PMD CLI with Java 22 - if I remember correctly, I only tested the build with Java 22 and eventually it works - that means the integration tests in pmd-dist were successful. However, I think we don't test the progress bar there.